src/Microsoft.IdentityModel.Tokens/ProvidersCache/HashCache.cs

```diff
@@ -0,0 +1,264 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.CryptoProviderCacheOptions
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.IdentityModel.Logging;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    internal class HashCache<TKey, TValue> : IProviderCache<TKey, TValue>
+    {
+        // delegates
+        internal delegate void ItemRemoved(TValue Value);
+        internal ItemRemoved OnItemRemoved { get; set; }
+
+        // _capacity is used for the newCacheSize calculation in the case where the cache is experiencing overflow
+        protected readonly int _capacity;
+
+        // the percentage of the cache to be removed when _maxCapacityPercentage is reached
+        protected readonly double _compactionPercentage = .20;
+
+        // When the current cache size gets to this percentage of _capacity, _compactionPercentage% of the cache will be removed.
+        protected readonly double _maxCapacityPercentage = .95;
```

Comment:
should these be configurable?

Reply:
@henrik-me , I think you are right, these values can be defined in the CryptoProviderCacheOptions. They are from the EventBasedLRUCache and seem reasonable default values.